common/asyncworkflow/queue/consumer/default_consumer.go

```diff
@@ -0,0 +1,262 @@
+) *DefaultConsumer {
+	ctx, cancelFn := context.WithCancel(context.Background())
+	c := &DefaultConsumer{
+		queueID:         queueID,
+		innerConsumer:   innerConsumer,
+		logger:          logger.WithTags(tag.AsyncWFQueueID(queueID)),
+		scope:           metricsClient.Scope(metrics.AsyncWorkflowConsumerScope),
+		frontendClient:  frontendClient,
+		ctx:             ctx,
+		cancelFn:        cancelFn,
+		shutdownTimeout: defaultShutdownTimeout,
+		msgDecoder:      codec.NewThriftRWEncoder(),
+		concurrency:     defaultConcurrency,
+	}
+
+	for _, opt := range options {
+		opt(c)
+	}
+
+	return c
+}
+
+func (c *DefaultConsumer) Start() error {
+	if err := c.innerConsumer.Start(); err != nil {
+		return err
+	}
+
+	for i := 0; i < c.concurrency; i++ {
+		c.wg.Add(1)
+		go c.runProcessLoop()
+		c.logger.Info("Started process loop", tag.Counter(i))
+	}
+	c.logger.Info("Started consumer", tag.Dynamic("concurrency", c.concurrency))
+	return nil
+}
+
+func (c *DefaultConsumer) Stop() {
+	c.logger.Info("Stopping consumer")
+	c.cancelFn()
+	c.wg.Wait()
+	if !common.AwaitWaitGroup(&c.wg, c.shutdownTimeout) {
+		c.logger.Warn("Consumer timed out on shutdown", tag.Dynamic("timeout", c.shutdownTimeout))
+		return
+	}
+
+	c.innerConsumer.Stop()
+	c.logger.Info("Stopped consumer")
+}
+
+func (c *DefaultConsumer) runProcessLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case msg, ok := <-c.innerConsumer.Messages():
+			if !ok {
+				c.logger.Info("Consumer channel closed")
+				return
+			}
+
+			c.processMessage(msg)
+		case <-c.ctx.Done():
+			c.logger.Info("Consumer context done so terminating loop")
+			return
+		}
+	}
+}
+
+func (c *DefaultConsumer) processMessage(msg messaging.Message) {
+	logger := c.logger.WithTags(tag.Dynamic("partition", msg.Partition()), tag.Dynamic("offset", msg.Offset()))
+	logger.Debug("Received message")
+
+	sw := c.scope.StartTimer(metrics.ESProcessorProcessMsgLatency)
+	defer sw.Stop()
+
+	var request sqlblobs.AsyncRequestMessage
+	if err := c.msgDecoder.Decode(msg.Value(), &request); err != nil {
+		logger.Error("Failed to decode message", tag.Error(err))
+		c.scope.IncCounter(metrics.AsyncWorkflowFailureCorruptMsgCount)
+		if err := msg.Nack(); err != nil {
+			logger.Error("Failed to nack message", tag.Error(err))
+		}
+		return
+	}
+
+	if err := c.processRequest(logger, &request); err != nil {
+		logger.Error("Failed to process message", tag.Error(err))
+		if err := msg.Nack(); err != nil {
+			logger.Error("Failed to nack message", tag.Error(err))
+		}
+		return
+	}
+
+	if err := msg.Ack(); err != nil {
+		logger.Error("Failed to ack message", tag.Error(err))
+	}
+	logger.Debug("Processed message successfully")
+}
+
+func (c *DefaultConsumer) processRequest(logger log.Logger, request *sqlblobs.AsyncRequestMessage) error {
+	switch request.GetType() {
+	case sqlblobs.AsyncRequestTypeStartWorkflowExecutionAsyncRequest:
+		startWFReq, err := decodeStartWorkflowRequest(request.GetPayload(), request.GetEncoding())
```

Comment:
After from payload, we also need to parse the headers and set the headers when sending the sync requests to frontend

Reply:
done